### PR TITLE
Do not limit the amount of gas on ETH transfers

### PR DESCRIPTION
**Pull request by fedgiac**
Remove gas limit when sending ETH.
This is done to be able to redeem claims using native ETH with wallets that don't support creating transactions that use access lists.

Some discussion on reentrancy. There are currently 5 exernal functions that act on the state (and don't revert): `claim`, `claimMany`, `swap`, `swapAll`, and `stopClaim` (see [here](https://github.com/gnosis/cow-token/blob/d13d7fdec82294e06ba35e742ac32d619e24c1a6/test/CowSwapVirtualToken.test.ts#L130-L159)).
Claims are set as claimed [before](https://github.com/gnosis/cow-token/blob/d13d7fdec82294e06ba35e742ac32d619e24c1a6/src/contracts/mixins/MerkleDistributor.sol#L188) actually performing the claim, so no claim can be executed twice. Calling any of the other functions, say `f`, in a reentrant transaction would be equivalent to calling `f` before the reentrant transaction, as the storage variables used by the reentrant claim transaction are not modified by `f` before the reentrant call is made.

This change does increase the cost of future changes to the code, as we should always keep into account the reentrancy risk when changing any of the mentioned functions.

### Test Plan

Modified old unit tests.

### CODE REVIEW

**Review by josojo — COMMENTED** [review]
Looks all good.

Not yet approving, as 
1) I don't understand why we need the PR. willl read upon in slack.

2) We don't have a test that shows the funds end up in the community fund. I think we have only the following tests:
```
UserOption
        ✓ withdraws WETH from the transaction executor to the communityFunds target (75ms)
        ✓ reverts if WETH withdrawal fails (118ms)
        ✓ reverts if sending too little ETH
        ✓ reverts if sending too much ETH
        ✓ reverts if ETH transfer fails (95ms)
        can use ETH for claiming
          ✓ executes successfully
          ✓ claims no tokens as instantly swappable
          ✓ increases total supply
          ✓ adds new cancelable vesting position
          ✓ emits token generation event
          ✓ can be claimed until immediately before the deadline
          ✓ cannot be claimed after deadline
          ````
         But we don't have a test: increases ETH balance  in community fund.

Sorry for being so picky.

**Comment by fedgiac** [comment]
All are good points!

1. [This](https://gnosisinc.slack.com/archives/C02RNR2SLKC/p1642028359076800) is a reference to the discussion. It seems that even just executing a user claim from Metamask requires some bad hack for it to work as expected, so since there are no security implications in doing this we decided it was easier to make our contract more permissive rather than use a hack and probably not being able to support other wallets.
2. We do test that a transfer is executed that sends tokens to the community funds target, which is namely:
```
✓ withdraws WETH from the transaction executor to the communityFunds target
```
Code is [here](https://github.com/gnosis/cow-token/blob/main/test/Claiming.test.ts#L404-L432), variables defined [here](https://github.com/gnosis/cow-token/blob/main/test/Claiming.test.ts#L571-L579). If the contract tried to send it to a different address, the test would fail as the mock would not be defined. The name of the test helps making sure that the target is the expected one. We do not test the balance with a real token implementation however, this is left to an e2e test and we currently didn't migrate e2e tests to this repo.
Something we don't check is that the community funds target does indeed receive the ETH from the user. I'll create a test for this.

File: src/contracts/mixins/Claiming.sol [hunk 1]
```diff
@@ -303,7 +305,20 @@ abstract contract Claiming is ClaimingInterface, VestingInterface, IERC20 {
...
+
+        // We transfer ETH using .call instead of .transfer as not to restrict
+        // the amount of gas sent to the target address during the transfer.
+        // This is particularly relevant for sending ETH to smart contracts:
+        // since EIP 2929, if a contract sends eth using `.transfer` then the
+        // transaction proposed to the node needs to specify an _access list_,
+        // which is currently not well supported by some wallet implementations.
+        // The drawback of this approach is that it requires the rest of the
+        // contract to be robust against reentrancy attacks at this point.
+        // solhint-disable-next-line avoid-low-level-calls
```
Comment (bh2smith) [inline src/contracts/mixins/Claiming.sol:317]: Very interesting story and good comment to put here. Well written too!

**Review by bh2smith — APPROVED** [review]

**Review by fleupold — COMMENTED** [review]
Change looks good, although I'm missing a discussion on the two alternative soltutions:

1. Limit the stipend to a value that's good enough for a Safe but still unlikely to allow re-entrancy (the argument in [slack](https://gnosisinc.slack.com/archives/C63LQTGNB/p1642070279031900) about future proofness is not super strong, as claiming this option will only work for 6 weeks in which no hardfork is scheduled).
2. Add a re-entrencry guard on `collectNativeTokenPayment`

I'm fine with the approach taken here (assuming re-review by Adam) but would like to hear your thoughts on the other 2 (which I personally feel may be less risky).

**Comment by fedgiac** [comment]
I'm against option 2 as:
1. it makes claims more expensive.
2. It's not obvious that adding a reentrancy guard to just `collectNativeTokenPayment` would fix the issue. In principle at this point you could call another function of the contract which changes the state, and this changed state is later read in the original transaction. So to be really reentrant-proof we should add a guard on any function that changes the state.

I think option 1 is definitely acceptable for us. On the other hand, I'm worried that restricting the gas to a specific amount might create a false sense of security: it's not reentrant because it's too expensive to do so. This is in general difficult to judge and I prefer that the full risk is fully exposed while discussing and on the audit. In principle reentrant contracts are not bad per se, but it's usually a bit of an overhead to check that reentrancy doesn't lead to exploiting a bad state of the contract.

**Review by josojo — APPROVED** [review]
Now I am convinced that it is good.

I think we don't need to change re-entrancy, because of:
1. Federico comment in PR is a good reasoning
(2. The "to" will be our safe that we control)

Usually, I prefer to not abstain from any longer reasoning, such as Federico's re-entrancy reasoning, in order to make the contract easier to understand, but I also see the disadvantages that Federico mentioned.

Proposal for improvement: 
Write the re-entrancy reasoning into the code as a comment such that the reasoning does not get lost.

**Review by nlordell — APPROVED** [review]
Code and reasoning seems solid to me, I can't think of any attacks.

That being said, I wouldn't be against just adding re-entrancy guards to all `external` methods in the vToken contract for piece of mind. While it does increase the cost of `claim`-ing and `swap`-ing these are operations that don't happen too often and might not be worth the extra attack surface given the contracts need a shelf-life of 4 years.

That being said, I can't see anything wrong with the reasoning now.

**Comment by fedgiac** [comment]
I ran the gas benchmarks on the current code and on the code with reentrancy guards ([this](https://github.com/gnosis/cow-token/compare/make-external-functions-non-reentrant)). On average, it seems that each call costs about 2500 gas more to be called, which is 1.12$ at current gas prices (3000$/ETH, 150 GWei/gas) per function call.

<details><summary>Results</summary>

```
|  Contract                       ·  Method    ·  Min  ·  Max  ·  Avg  · # calls │
== With reentrancy guard ==
|  MerkleDistributorTestInterface ·  claim     · 40472 · 61120 · 43967 ·    609  │
|  MerkleDistributorTestInterface ·  claimMany · 62289 · 79429 · 71450 ·     12  │
|  ClaimingTestInterface          ·  swap      · 49047 · 49071 · 49056 ·      8  │
|  ClaimingTestInterface          ·  swapAll   · 41310 · 44037 · 41651 ·      8  │
== Without reentrancy guard ==
|  MerkleDistributorTestInterface ·  claim     · 38086 · 58734 · 41581 ·    609  │
|  MerkleDistributorTestInterface ·  claimMany · 59900 · 76944 · 69005 ·     12  │
|  ClaimingTestInterface          ·  swap      · 46709 · 46733 · 46718 ·      8  │
|  ClaimingTestInterface          ·  swapAll   · 37200 · 41699 · 37762 ·      8  │
```
</details>

>  Write the re-entrancy reasoning into the code as a comment such that the reasoning does not get lost.

The problem with writing this reasoning down is that it will become outdated quicky if the code changes significantly.

On the other hand, my assumption is that the token code is not going to change significantly at this point. Otherwise, I'd have proposed to use an eth collector contract that receives eth and can be queried at any time by the actual owner for withdrawing the ETH.

**Review by fleupold — APPROVED** [review]
Thanks for the explanation. I would slightly prefer capturing it in the code (also for auditors to be aware that we thought about it), but not a must.

**Pull request merged**